platform/posix/transport/include/mbedtls_pkcs11_posix.h

```diff
@@ -0,0 +1,205 @@
+    mbedtls_pk_context privKey;           /**< @brief Client private key context. */
+    mbedtls_pk_info_t privKeyInfo;        /**< @brief Client private key info. */
+
+    /* PKCS #11. */
+    CK_FUNCTION_LIST_PTR pP11FunctionList;
+    CK_SESSION_HANDLE p11Session;
+    CK_OBJECT_HANDLE p11PrivateKey;
+    CK_KEY_TYPE keyType;
+} MbedtlsPkcs11Context_t;
+
+/**
+ * @brief TLS Connect / Disconnect return status.
+ */
+typedef enum MbedtlsPkcs11Status
+{
+    MBEDTLS_PKCS11_SUCCESS = 0,         /**< Function successfully completed. */
+    MBEDTLS_PKCS11_INVALID_PARAMETER,   /**< At least one parameter was invalid. */
+    MBEDTLS_PKCS11_INSUFFICIENT_MEMORY, /**< Insufficient memory required to establish connection. */
+    MBEDTLS_PKCS11_INVALID_CREDENTIALS, /**< Provided credentials were invalid. */
+    MBEDTLS_PKCS11_HANDSHAKE_FAILED,    /**< Performing TLS handshake with server failed. */
+    MBEDTLS_PKCS11_INTERNAL_ERROR,      /**< A call to a system API resulted in an internal error. */
+    MBEDTLS_PKCS11_CONNECT_FAILURE      /**< Initial connection to the server failed. */
+} MbedtlsPkcs11Status_t;
+
+/**
+ * @brief Contains the credentials necessary for tls connection setup.
+ */
+typedef struct MbedtlsPkcs11Credentials
+{
+    /**
+     * @brief To use ALPN, set this to a NULL-terminated list of supported
+     * protocols in decreasing order of preference.
+     *
+     * See [this link]
+     * (https://aws.amazon.com/blogs/iot/mqtt-with-tls-client-authentication-on-port-443-why-it-is-useful-and-how-it-works/)
+     * for more information.
+     */
+    const char ** pAlpnProtos;
+
+    /**
+     * @brief Disable server name indication (SNI) for a TLS session.
+     */
+    bool disableSni;
+
+    const char * pRootCaPath;     /**< @brief String representing a trusted server root certificate. */
+    char * pClientCertLabel;      /**< @brief String representing the PKCS #11 label for the client certificate. */
+    char * pPrivateKeyLabel;      /**< @brief String representing the PKCS #11 label for the private key. */
+    CK_SESSION_HANDLE p11Session; /**< @brief PKCS #11 session handle. */
+} MbedtlsPkcs11Credentials_t;
+
+/**
+ * @brief Sets up a TLS session on top of a TCP connection using the MbedTLS API.
+ *
+ * @param[out] pNetworkContext The output parameter to return the created network context.
+ * @param[in] pHostName The hostname of the remote endpoint.
+ * @param[in] port The destination port.
+ * @param[in] pMbedtlsPkcs11Credentials Credentials for the TLS connection.
+ *
+ * @note A timeout of 0 means infinite timeout.
+ *
+ * @return #MBEDTLS_PKCS11_SUCCESS on success;
+ * #MBEDTLS_PKCS11_INSUFFICIENT_MEMORY, #MBEDTLS_PKCS11_INVALID_CREDENTIALS,
+ * #MBEDTLS_PKCS11_HANDSHAKE_FAILED, #MBEDTLS_PKCS11_INTERNAL_ERROR,
+ * or #MBEDTLS_PKCS11_CONNECT_FAILURE on failure.
+ */
+MbedtlsPkcs11Status_t Mbedtls_Pkcs11_Connect( NetworkContext_t * pNetworkContext,
```

Comment:
We have discussed in the past that the Connect/Disconnect calls do not need to use `NetworkContext_t` input type as they are not part of the transport interface, 
```suggestion
MbedtlsPkcs11Status_t Mbedtls_Pkcs11_Connect( MbedtlsPkcs11Context_t * pSslContext,
```